### PR TITLE
gateway: Support for custom endpoint.

### LINKED ISSUE #4083: gateway: Support for custom endpoint
Support endpoint in the format:

minio gateway azure [endpoint]
default would connect to Azure backend.

minio gateway s3 [endpoint]
default would connect to AWS-S3.

### PR DESCRIPTION
**Pull request by krishnasrinivas**
fixes #4083

### CODE REVIEW

File: cmd/gateway-main.go [hunk 1]
```diff
@@ -142,10 +142,14 @@ func gatewayMain(ctx *cli.Context) {
 	// First argument is selected backend type.
 	backendType := ctx.Args().First()
 
+	// Second argument is endpoint.	If no endpoint is specified then the
+	// gateway implementation should use a default setting.
+	endPoint := ctx.Args().Get(1)
```
Comment (harshavardhana) [inline cmd/gateway-main.go]: is there no input validation? - I think we should.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/gateway-main.go [hunk 2]
```diff
@@ -117,6 +118,25 @@ func newGatewayConfig(accessKey, secretKey, region string) error {
 	return nil
 }
```
Comment (harshavardhana) [inline cmd/gateway-main.go:121]: perhaps a better name is `parseGatewayEndpoint`  and also a comment as well.

File: cmd/gateway-main.go [hunk 3]
```diff
@@ -117,6 +118,25 @@ func newGatewayConfig(accessKey, secretKey, region string) error {
...
+		return "", false, err
+	}
+
+	switch u.Scheme {
+	case "http":
+		return u.Host, true, nil
+	case "https":
+		return u.Host, false, nil
+	case "":
+		// url.Parse("play.minio.io") puts play.minio.io in u.Path
```
Comment (harshavardhana) [inline cmd/gateway-main.go]: In go1.8 this wouldn't happen anymore and an error will be generated, if we need to support a scenario like this. We should handle it like `net.SplitHostPort()` style.

Comment (krishnasrinivas) [reply cmd/gateway-main.go]: but it seems to work:
```
krishna@escape:~/tmp$ cat url.go
package main

import (
	"fmt"
	"net/url"
)

var asd = "asd"

func main() {
	u, err := url.Parse("play.minio.io")
	fmt.Println(err, u.Path)
}
krishna@escape:~/tmp$ go version
go version go1.8.1 linux/amd64
krishna@escape:~/tmp$ go run url.go
<nil> play.minio.io
krishna@escape:~/tmp$ 

```

Comment (harshavardhana) [reply cmd/gateway-main.go]: Doesn't work with IP @krishnasrinivas 

```
$ go run url.go
parse 192.168.1.106:9000: first path segment in URL cannot contain colon <nil>
```

File: cmd/gateway-main.go [hunk 4]
```diff
@@ -117,6 +119,31 @@ func newGatewayConfig(accessKey, secretKey, region string) error {
 	return nil
 }
 
+// Return endpoint.
+func parseGatewayEndpoint(arg string) (endPoint string, secure bool, err error) {
```
Comment (harshavardhana) [inline cmd/gateway-main.go:123]: add a test case? @krishnasrinivas

File: cmd/gateway-main.go [hunk 5]
```diff
@@ -117,6 +119,31 @@ func newGatewayConfig(accessKey, secretKey, region string) error {
...
+	schemeSpecified, err := regexp.MatchString(".*://", endPoint)
```
Comment (donatello) [inline cmd/gateway-main.go]: `arg` instead of `endPoint`? `endPoint` is part of return and at this point it is an empty string.

Comment (donatello) [reply cmd/gateway-main.go]: Also, for a simple check like this, it is not necessary to use regular expressions:

```
schemeSpecified := strings.Contains(arg, "://")
```

This is faster, and there is no error checking needed.

**Review by donatello — CHANGES_REQUESTED** [review]

**Comment by harshavardhana** [comment]
can you add a test case @krishnasrinivas ?

**Review by harshavardhana — APPROVED** [review]

**Comment by harshavardhana** [comment]
@krishnasrinivas error in tests 

```
Installing ineffassign: Done.
Running vet: cmd/gateway-main_test.go:45: wrong number of args for format in Errorf call: 7 needed but 8 args
make: *** [vet] Error 1
```

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**